mir_eval/transcription_velocity.py

```diff
@@ -0,0 +1,354 @@
+"""
+Transcription evaluation, as defined in :mod:`mir_eval.transcription`, does not
+take into account the velocities of reference and estimated notes. This
+submodule implements a variant of
+:func:`mir_eval.transcription.precision_recall_f1_overlap` which
+additionally considers note velocity when determining whether a note is
+correctly transcribed. This is done by defining a new function
+:func:`mir_eval.transcription_velocity.match_notes` which first calls
+:func:`mir_eval.transcription.match_notes` to get a note matching based on
+onset, offset, and pitch. Then, we follow the evaluation procedure described in
+[#hawthorne2018onsets]_ to test whether an estimated note should be considered
+correct:
+
+1. Reference velocities are re-scaled to the range [0, 1].
+
+2. A linear regression is performed to estimate global scale and offset
+   parameters which minimize the L2 distance between matched estimated and
+   (rescaled) reference notes.
+
+3. The scale and offset parameters are used to rescale estimated
+   velocities.
+
+4. An estimated/reference note pair which has been matched according to the
+   onset, offset, and pitch is further only considered correct if the rescaled
+   velocities are within a predefined threshold, defaulting to 0.1.
+
+:func:`mir_eval.transcription_velocity.match_notes` is used to define a new
+variant :func:`mir_eval.transcription_velocity.precision_recall_f1_overlap`
+which considers velocity.
+
+Conventions
+-----------
+
+This submodule follows the conventions of :mod:`mir_eval.transcription` and
+additionally requires velocities to be provided as MIDI velocities in the range
+[0, 127].
+
```

Comment:
validation code does not seem to enforce the upper bound?

Reply:
True - it's a bit odd, because the only reason to specify MIDI velocities is so that this line makes sense:
> `velocity_range = max(1, max_velocity - min_velocity)`
It's ok if `max_velocity - min_velocity = 0` because that just means all the velocities in the reference annotation are the same velocity, but if `0 < max_velocity - min_velocity < 1` then it's a bit odd to clip it to 1.  Instead of picking some epsilon, I thought it would be simpler to just specify velocity should be MIDI velocity.  It actually doesn't matter if the velocity range is [0, 127] or [0, N] for some large N, since the velocity range will just get rescaled to `[0, 1]` anyways.  So there's not really any reason to check for `< 127`, if that makes sense.

Reply:
That seems reasonable.  I think the docstring should be amended to state the actual assumptions of the code though.

Reply:
What other assumptions do you think should be listed?

Reply:
If it doesn't require that velocity is <127, why not just say that it has to be a non-negative number?

Reply:
Because 
> if 0 < max_velocity - min_velocity < 1 then it's a bit odd to clip it to 1.

So non-negative is not really sufficient to describe the intended range.  I can't really think of a helpful concrete way to describe it, like "velocity should be non-negative, but make sure the range is pretty big, probably you should just use MIDI velocities but if you use some other large velocity range you'll probably be ok as long as the min - max is greater than 1".